### PR TITLE
GH-3245 Change password from settings page

### LINKED ISSUE #3245: Change password from settings
# Expected behavior

### 1. Go to profile settings

In the `Profile` settings, we would like to add a new section that allows users to reset their password:

image
https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=13273-96299&mode=design&t=tY8CUvxiUNT2VRnR-11

### 2. Receive an email 

Selecting `Change password` will trigger an email to the user with a reset link:

Email body style & content:

image
https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=16594-92960&mode=design&t=tY8CUvxiUNT2VRnR-11

### 3. Change password

When the user clicks on the email link, they should be redirected to a page at `app.twenty.com/reset-password`. This page will prompt them to enter a new password:

image
https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=19100-70339&mode=design&t=tY8CUvxiUNT2VRnR-11

### PR DESCRIPTION
**Pull request by i-am-chitti**
## Description
This PR adds password reset feature to change password from settings page. Here is the workflow -

1. On clicking "Change Password", a reset token is generated. This is done via `emailPasswordResetLink` mutation. It generates the token and store its hash in DB with an expiry. It sends back the confirmation to user. 
2. It mails the user their plain token. If user tries to generate the token again within a certain time defined in constant `PASSWORD_RESET_TOKEN_EXPIRES_IN`, it will throw error. Also, this constant is optional. Its default is 5 minutes.
3. Once user opens the link present in the email, he will be directed to `reset-password/:token` path. Here a validation of token is carried out through `validatePasswordResetToken` query. If invalid, in case of non-logged in, redirected to sign in page else to index page.
4. Next, password reset form appears. He enters the new password. Updation is done via mutation `updatePasswordViaResetToken`. It validates the token again and then, new password. New password should be same as old. On success, it updates the password and sends the success confirmation.
5. Now, if user was logged in, he will be directed to sign in page. Else, he is logged in programatically and directed to sign in page.

Have a look at the attached Gdrive link where complete flow has been recorded.

### Issue
- #3245 

### Note
- Migration has been already added to add the DB column for storing token and expiry.
- There was some unknown error during graphql data generation. Hence, had to use `skipValidation` in `codegen.cjs`. This error is not related to mine changes. It started popping once I synced the main branch. **Request you to kindly regenerate the graphql files before merging**.
image

### Screenshots
image

image

image

image

image

### Screencast

https://drive.google.com/file/d/1y6q6ocakQa6sScMuPZQYdDeXGuuNyFyQ/view?usp=sharing

### CODE REVIEW

**Comment by charlesBochet** [comment]
@martmull could you check the backend side?
It would also be great to have a review from @thaisguigon or @lucasbordeau regarding the FE side!

File: packages/twenty-server/src/core/auth/services/token.service.ts
Comment (martmull) [inline packages/twenty-server/src/core/auth/services/token.service.ts:378]:
```suggestion
  async sendEmailPasswordResetLink(
```

**Review by martmull — COMMENTED** [review]
Hey @i-am-chitti, it is really good work. I have left one minor comment. I could not try your code because of the graphql schema issue. I will investigate that and come back to you

**Comment by martmull** [comment]
Hey @i-am-chitti, there is a PR to move email templating in a dedicated package (https://github.com/twentyhq/twenty/pull/3542). Do you want me to wait for your PR to be merged?

**Comment by i-am-chitti** [comment]
Hey @martmull, looks like this PR will be under review for sometime as there is some CI builds failing. If your PR is ready, you can go ahead with merge. In both cases, I'll need to update the email template changes here as well as the email module has been moved to dedicated package.

**Comment by martmull** [comment]
@i-am-chitti ok, I will merge, and then explain the new email templating structure for you to adapt your code. Thank you!

**Comment by martmull** [comment]
@i-am-chitti So now email templates are in packages/twenty-emails. You can merge main in your branch and launch `yarn` to update dependancies. Move your templates (resetPasswordLink and resetPasswordNotification)  in `packages/twenty-emails/src/emails`. Then launch `yarn nx run twenty-emails:build` to build your email package again, the added templates should be available in `packages/twenty-server` like in `packages/twenty-server/src/workspace/cron/clean-inactive-workspaces/clean-inactive-workspace.job.ts`:
```ts
...
import {
  CleanInactiveWorkspaceEmail,
  DeleteInactiveWorkspaceEmail,
} from 'twenty-emails';
...
```
Do not hesitate to ping me if you encounter any issue

**Comment by martmull** [comment]
@i-am-chitti merging main in your branch should fix the `graphql:data:generate` command also

File: packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx
Comment (martmull) [inline packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx:23]: missing 🗝 
image
```suggestion
            <Title value="Reset your password 🗝" />
```

File: packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx
Comment (martmull) [inline packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx:27]: helloString can be removed (see figma)

File: packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx
Comment (martmull) [inline packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx:31]: Think we will need a Custom EmailLink component for twenty emails to match Figma design (we should have gray links not blue links) 
image

File: packages/twenty-server/src/core/auth/emails/password-update-notify.email.tsx
Comment (martmull) [inline packages/twenty-server/src/core/auth/emails/password-update-notify.email.tsx:28]: > don't see any Figma design for that mail, did you infer it on your onw? If yes I will validate the design and > wording with @Bonapara 

edit: Seen with @Bonapara here is a Figma design for this email https://www.figma.com/file/xt8O9mFeLl46C5InWwoMrN/Twenty?type=design&node-id=20829-56525&mode=design&t=yclxfsOGGY9rOqAL-0
The Object would be: `Your Password Has Been Successfully Changed`
image

**Review by martmull — CHANGES_REQUESTED** [review]

**Comment by i-am-chitti** [comment]
Thanks @martmull for the review. I'll be pushing the updates soon.

File: packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx
Comment (i-am-chitti) [reply packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx]: Fixed, see the attached screenshot in PR description

File: packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx
Comment (i-am-chitti) [reply packages/twenty-server/src/core/auth/emails/password-reset-link.email.tsx]: Have created a custom component for the link.

File: packages/twenty-server/src/core/auth/services/token.service.ts
Comment (i-am-chitti) [reply packages/twenty-server/src/core/auth/services/token.service.ts]: Updated the method name.

File: packages/twenty-server/src/core/auth/emails/password-update-notify.email.tsx
Comment (i-am-chitti) [reply packages/twenty-server/src/core/auth/emails/password-update-notify.email.tsx]: I didn't find this design in the original issue. Thanks for noting it. I've made necessary changes and now, it should match with the Figma one.

**Comment by i-am-chitti** [comment]
@martmull, I've made the changes. You can have a re-review.

File: packages/twenty-emails/src/emails/password-reset-link.email.tsx
Comment (martmull) [inline packages/twenty-emails/src/emails/password-reset-link.email.tsx:16]: `userName` not used anymore

File: packages/twenty-emails/src/emails/password-reset-link.email.tsx
Comment (martmull) [inline packages/twenty-emails/src/emails/password-reset-link.email.tsx:30]: looks like tabulation is 4 spaces when it should be 2 spaces

File: packages/twenty-emails/src/emails/password-update-notify.email.tsx
Comment (martmull) [inline packages/twenty-emails/src/emails/password-update-notify.email.tsx:33]: tabulation idem

File: packages/twenty-emails/tsup.index.tsx
Comment (martmull) [inline packages/twenty-emails/tsup.index.tsx:4]: new line missing

**Review by martmull — CHANGES_REQUESTED** [review]
Hey @i-am-chitti, that's really good work. I have some minor changes regarding email css and code style. I don't want to be too touchy on that so don't hesitate to ask me to pass those changes! Also, I will ask @lucasbordeau to double check the front part. Then we will be able to merge 🎉. Really nice job again!

File: packages/twenty-emails/src/emails/password-reset-link.email.tsx
Comment (lucasbordeau) [reply packages/twenty-emails/src/emails/password-reset-link.email.tsx:30]: We should add the same eslint config as in twenty-front package

**Comment by lucasbordeau** [comment]
Nice work !

**Comment by martmull** [comment]
@i-am-chitti , this diff will fix your tests also 
image

File: packages/twenty-emails/tsup.index.tsx
Comment (i-am-chitti) [reply packages/twenty-emails/tsup.index.tsx]: Not able to put new line, it was being removed on saving the file in VSCode. Can we skip this?

**Comment by i-am-chitti** [comment]
@martmull Thanks for the diff and helping hand. Indeed, feel free to push the changes in PR itself. 
Thanks @lucasbordeau for the review. I've pushed changes for PR feedbacks.

I've pushed the diff, but seems there are further tests failing related to emails. Could you have a look once and direct me how to get them fixed?

File: packages/twenty-emails/tsup.index.tsx
Comment (martmull) [reply packages/twenty-emails/tsup.index.tsx]: yes np

**Comment by martmull** [comment]
@i-am-chitti thank you for your changes. I have added one last commit to fix the CI (it was an issue with twenty-emails package, not your fault). Think we are able to merge now 🚀. Thank you very much for your contrib, it is really good job. Cheers  🤗

**Review by martmull — APPROVED** [review]

**Comment by FelixMalfait** [comment]
Great work @i-am-chitti !!! Thanks!

**Pull request merged**
